dashboard/app/models/foorm/library_question.rb

```diff
@@ -42,10 +45,12 @@ def published_forms_appeared_in
     end
   end
 
-  # TO DO: names need to be tracked in the question column (and JSON) to make sure libraries continue to be valid SurveyJS.
-  # TO DO: check that the library question has the same name in the question as in the question_name field.
-  # TO DO: check that the library question name is unique within the library (there is DB validation for this)
   def validate_library_question
+    # Keep question name stored in the question JSON field in sync with what's in the database
+    if JSON.parse(question)['name'] != question_name
+      raise InvalidFoormConfigurationError, 'library question name in question JSON must match name of library question name in database'
```

Comment:
Is this error displayed on the frontend somehow?

Reply:
Yes -- it's supposed to show an error if a save is attempted, like this (yikes on the formatting...):

![image](https://user-images.githubusercontent.com/25372625/115093357-df324f80-9ece-11eb-94a7-40bf8e61fe8b.png)

The pattern of having a named error class is [stolen from the Form model](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/foorm/form.rb#L22).

That said, your question raises another problem -- we have a "Validate" button on the editor (visible in screenshot ^^^) that checks the content of the library question you're editing and tells you if it's valid. With this change as-is, all library questions are invalid, because we have a little controller method that validates a new Library Question by building [a new object with only the JSON contents of what is being edited](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/api/v1/pd/foorm/library_questions_controller.rb#L6). Since the Library Question doesn't have a name, it's always invalid :(.

I'm realizing that Forms have this this same problem (ie, an editor could update the name of the Form in the JSON, and, once saved, the name in the attribute `form_name` would be different from what is in the JSON.

I think I'm going to have to circle back a little bit with Tess/other PLC folks to figure out exactly how we want this to work...

(sorry if this is a stream of consciousness -- helpful for me to write it out, and hopefully at least some of it is coherent...)

Reply:
To fix the validate button problem, I've added a change to have the library question validation run in the context of the existing library question (if one exists). That way, we can compare the updated JSON and make sure that the name matches.